Some tests are flaky

> I'm building it (with `nix build`, I should have tried without because it's faster) and after 4 times the 5th one it failed:
> 
> <details>
> 
> ```ncps> --- FAIL: TestGetNar (2.05s)
> ncps>     cache_test.go:889: /nix/store/g1z2c7x7mj9za56yxcg4qmrdhnp91hfg-dbmate-2.27.0/bin/dbmate --url=sqlite:/build/cache-path-3920495635/var/ncps/db/db.sqlite --migrations-dir=/build/source/db/migrations --schema-file=/build/source/db/schema.sql up: Creating: /build/cache-path-3920495635/var/ncps/db/db.sqlite
> ncps>         Applying: 20241210054814_create-narinfos-table.sql
> ncps>         Applied: 20241210054814_create-narinfos-table.sql in 2.570205ms
> ncps>         Applying: 20241210054829_create-nars-table.sql
> ncps>         Applied: 20241210054829_create-nars-table.sql in 3.441667ms
> ncps>         Applying: 20241213014846_add-query-to-nars-table.sql
> ncps>         Applied: 20241213014846_add-query-to-nars-table.sql in 2.512475ms
> ncps>     --- FAIL: TestGetNar/nar_exists_upstream (2.03s)
> ncps>         --- FAIL: TestGetNar/nar_exists_upstream/nar_does_exist_in_the_database,_and_has_initial_last_accessed_at (0.00s)
> ncps>             cache_test.go:1026:
> ncps>                   Error Trace:    /build/source/pkg/cache/cache_test.go:1026
> ncps>                   Error:          Not equal:
> ncps>                                   expected: time.Date(2025, time.September, 24, 18, 25, 15, 0, time.UTC)
> ncps>                                   actual  : time.Date(2025, time.September, 24, 18, 25, 16, 0, time.UTC)
> ncps> 
> ncps>                                   Diff:
> ncps>                                   --- Expected
> ncps>                                   +++ Actual
> ncps>                                   @@ -1,2 +1,2 @@
> ncps>                                   -(time.Time) 2025-09-24 18:25:15 +0000 UTC
> ncps>                                   +(time.Time) 2025-09-24 18:25:16 +0000 UTC
> ncps> 
> ncps>                   Test:           TestGetNar/nar_exists_upstream/nar_does_exist_in_the_database,_and_has_initial_last_accessed_at
> ncps>         --- FAIL: TestGetNar/nar_exists_upstream/pulling_it_another_time_within_recordAgeIgnoreTouch_should_not_update_last_accessed_at (1.00s)
> ncps>             --- FAIL: TestGetNar/nar_exists_upstream/pulling_it_another_time_within_recordAgeIgnoreTouch_should_not_update_last_accessed_at/narinfo_does_exist_in_the_database_with_the_same_last_accessed_at (0.00s)
> ncps>                 cache_test.go:1072:
> ncps>                           Error Trace:    /build/source/pkg/cache/cache_test.go:1072
> ncps>                           Error:          Not equal:
> ncps>                                           expected: time.Date(2025, time.September, 24, 18, 25, 15, 0, time.UTC)
> ncps>                                           actual  : time.Date(2025, time.September, 24, 18, 25, 16, 0, time.UTC)
> ncps> 
> ncps>                                           Diff:
> ncps>                                           --- Expected
> ncps>                                           +++ Actual
> ncps>                                           @@ -1,2 +1,2 @@
> ncps>                                           -(time.Time) 2025-09-24 18:25:15 +0000 UTC
> ncps>                                           +(time.Time) 2025-09-24 18:25:16 +0000 UTC
> ncps> 
> ncps>                           Test:           TestGetNar/nar_exists_upstream/pulling_it_another_time_within_recordAgeIgnoreTouch_should_not_update_last_accessed_at/narinfo_does_exist_in_the_database_with_the_same_last_accessed_at
> ncps> FAIL
> ncps> FAIL      github.com/kalbasit/ncps/pkg/cache      5.675s
> ncps> FAIL
> error: Cannot build '/nix/store/zrypa8iw0ybqkr4siyczhj6xv7d7sxny-ncps-6a74f92.drv'.
>        Reason: builder failed with exit code 1.
>        Output paths:
>          /nix/store/0rdgxs023ngh8yimkcdppihdqiir754n-ncps-6a74f92
>        Last 25 log lines:
>        >                                        +++ Actual
>        >                                      @@ -1,2 +1,2 @@
>        >                                         -(time.Time) 2025-09-24 18:25:15 +0000 UTC
>        >                                      +(time.Time) 2025-09-24 18:25:16 +0000 UTC
>        >
>        >                      Test:           TestGetNar/nar_exists_upstream/nar_does_exist_in_the_database,_and_has_initial_last_accessed_at
>        >         --- FAIL: TestGetNar/nar_exists_upstream/pulling_it_another_time_within_recordAgeIgnoreTouch_should_not_update_last_accessed_at (1.00s)
>        >             --- FAIL: TestGetNar/nar_exists_upstream/pulling_it_another_time_within_recordAgeIgnoreTouch_should_not_update_last_accessed_at/narinfo_does_exist_in_the_database_with_the_same_last_accessed_at (0.00s)
>        >                 cache_test.go:1072:
>        >                       Error Trace:    /build/source/pkg/cache/cache_test.go:1072
>        >                      Error:          Not equal:
>        >                                      expected: time.Date(2025, time.September, 24, 18, 25, 15, 0, time.UTC)
>        >                                          actual  : time.Date(2025, time.September, 24, 18, 25, 16, 0, time.UTC)
>        >
>        >                                          Diff:
>        >                                           --- Expected
>        >                                            +++ Actual
>        >                                      @@ -1,2 +1,2 @@
>        >                                         -(time.Time) 2025-09-24 18:25:15 +0000 UTC
>        >                                      +(time.Time) 2025-09-24 18:25:16 +0000 UTC
>        >
>        >                      Test:           TestGetNar/nar_exists_upstream/pulling_it_another_time_within_recordAgeIgnoreTouch_should_not_update_last_accessed_at/narinfo_does_exist_in_the_database_with_the_same_last_accessed_at
>        > FAIL
>        > FAIL       github.com/kalbasit/ncps/pkg/cache      5.675s
>        > FAIL
>        For full logs, run:
>          nix log /nix/store/zrypa8iw0ybqkr4siyczhj6xv7d7sxny-ncps-6a74f92.drv
> ```
> </details> 

 _Originally posted by @aciceri in [#314](https://github.com/kalbasit/ncps/issues/314#issuecomment-3330165228)_

--- a/pkg/cache/cache_test.go
+++ b/pkg/cache/cache_test.go
@@ -940,36 +940,18 @@ func TestGetNar(t *testing.T) {
 
 		nu := nar.URL{Hash: testdata.Nar1.NarHash, Compression: nar.CompressionTypeXz}
 
-		var size int64
-
-		var r io.ReadCloser
-
-		for i := 1; i < 100; i++ {
-			// NOTE: I tried runtime.Gosched() but it makes the test flaky
-			time.Sleep(time.Duration(i) * time.Millisecond)
-
-			var err error
-			size, r, err = c.GetNar(context.Background(), nu)
+		t.Run("able to get the NAR even in flight from upstream", func(t *testing.T) {
+			_, r, err := c.GetNar(context.Background(), nu)
 			require.NoError(t, err)
 
-			if size > 0 {
-				break
-			}
-		}
+			t.Run("body is the same", func(t *testing.T) {
+				body, err := io.ReadAll(r)
+				require.NoError(t, err)
 
-		defer r.Close()
-
-		t.Run("size is correct", func(t *testing.T) {
-			assert.Equal(t, int64(len(testdata.Nar1.NarText)), size)
-		})
-
-		t.Run("body is the same", func(t *testing.T) {
-			body, err := io.ReadAll(r)
-			require.NoError(t, err)
-
-			if assert.Len(t, testdata.Nar1.NarText, len(string(body))) {
-				assert.Equal(t, testdata.Nar1.NarText, string(body))
-			}
+				if assert.Len(t, testdata.Nar1.NarText, len(string(body))) {
+					assert.Equal(t, testdata.Nar1.NarText, string(body))
+				}
+			})
 		})
 
 		t.Run("it should now exist in the store", func(t *testing.T) {
@@ -985,11 +967,6 @@ func TestGetNar(t *testing.T) {
 				}
 			}
 
-			assert.NoError(t, err)
-		})
-
-		t.Run("getting the narinfo so the record in the database now exists", func(t *testing.T) {
-			_, err := c.GetNarInfo(context.Background(), testdata.Nar1.NarInfoHash)
 			assert.NoError(t, err)
 		})
 
@@ -1076,9 +1053,13 @@ func TestGetNar(t *testing.T) {
 
 			nu := nar.URL{Hash: testdata.Nar1.NarHash, Compression: nar.CompressionTypeXz}
 
-			_, r, err := c.GetNar(context.Background(), nu)
+			size, r, err := c.GetNar(context.Background(), nu)
 			require.NoError(t, err)
 			defer r.Close()
+
+			t.Run("size is correct", func(t *testing.T) {
+				assert.Equal(t, int64(len(testdata.Nar1.NarText)), size)
+			})
 
 			t.Run("narinfo does exist in the database, and has more recent last_accessed_at", func(t *testing.T) {
 				const query = `

